webapp/javascript/components/TimelineChart/Annotations.plugin.tsx

```diff
@@ -0,0 +1,285 @@
+  const plotOffset = plot.getPlotOffset() as {
+    top: number;
+    left: number;
+  };
+
+  return {
+    x: Math.floor(extractedX.axis.p2c(positionInTimestamp.x)) + plotOffset.left,
+    y: Math.floor(extractedY.axis.p2c(positionInTimestamp.y)) + plotOffset.top,
+  };
+};
+
+const findAnnotationByCursorPosition = (
+  x: number,
+  y: number,
+  list: AnnotationPosition[] = []
+) => {
+  return list?.find((an) => {
+    return x >= an.fromX && x <= an.toX && y >= an.fromY && y <= an.toY;
+  });
+};
+
+function drawRoundRect(
+  ctx: CtxType,
+  x: number,
+  y: number,
+  w: number,
+  h: number,
+  radius: number,
+  color: string | Color
+) {
+  const r = x + w;
+  const b = y + h;
+  ctx.beginPath();
+  ctx.strokeStyle = color;
+  ctx.fillStyle = color;
+  ctx.lineWidth = 1;
+  ctx.moveTo(x + radius, y);
+  ctx.lineTo(r - radius, y);
+  ctx.quadraticCurveTo(r, y, r, y + radius);
+  ctx.lineTo(r, y + h - radius);
+  ctx.quadraticCurveTo(r, b, r - radius, b);
+  ctx.lineTo(x + radius, b);
+  ctx.quadraticCurveTo(x, b, x, b - radius);
+  ctx.lineTo(x, y + radius);
+  ctx.quadraticCurveTo(x, y, x + radius, y);
+  ctx.fill();
+  ctx.stroke();
+}
+
+(function ($) {
+  function init(plot: jquery.flot.plot & jquery.flot.plotOptions) {
+    const annotationsPositions: AnnotationPosition[] = [];
+    const placeholder = plot.getPlaceholder();
+
+    function onHover(_: unknown, pos: { x: number; y: number }) {
+      if (annotationsPositions?.length) {
+        const { x, y } = getCursorPositionInPx(plot, pos);
+
+        const annotation = findAnnotationByCursorPosition(
+          x,
+          y,
+          annotationsPositions
+        );
+
+        if (annotation) {
+          placeholder.trigger('hoveringOnAnnotation', [{ hovering: true }]);
+        } else {
+          placeholder.trigger('hoveringOnAnnotation', [{ hovering: false }]);
+        }
+      }
+    }
+
+    function onClick(
+      _: unknown,
+      pos: { x: number; pageX: number; pageY: number; y: number }
+    ) {
+      const options: IFlotOptions = plot.getOptions();
+      const container = inject($);
+      const containerEl = container?.[0];
+
+      ReactDOM.unmountComponentAtNode(containerEl);
+
+      const ContextMenu = options?.ContextMenu;
+
+      const { x, y } = getCursorPositionInPx(plot, pos);
+
+      const annotation = findAnnotationByCursorPosition(
+        x,
+        y,
+        annotationsPositions
+      );
+
+      if (ContextMenu && containerEl) {
+        const timestamp = Math.round(pos.x / 1000);
+        const value = annotation
+          ? {
+              timestamp: annotation.timestamp,
+              content: annotation.content,
+            }
+          : null;
+
+        ReactDOM.render(
+          <ContextMenu
+            click={{ ...pos }}
+            containerEl={containerEl}
+            timestamp={timestamp}
+            value={value}
+          />,
+          containerEl
+        );
```

Comment:
I think we can organize this better...Annotations plugin should not be responsible for rendering a Context Menu... Annotations plugin should only be responsible for rendering Annotations.

Reply:
can you please explain a bit more. This function opens context menu on click on canvas. Maybe I have to move this outside of flot's plugin as function?